TSA lesson-based doc on SDLC (#902)
digital-asset/docs.daml.com: docs/3.1/docs/tsa-arch/sdlc.rst

Daml Application Lifecycle: Considerations and Best Practices
#############################################################
The Software Development Lifecycle (SDLC) of a Daml application mirrors that of a Web2 or enterprise service with facets such as:

* Design
* Develop
* Test
* Deploy
* Operations Support

> However, Daml applications differ significantly in architecture, as cross-organizational workflows and state are explicitly defined in Daml code. This requires specific considerations during each phase to account for these differences. 

Reviewer: We should choose either "phase" or "facet" and use it consistently. My preference would be "phase", as I think it's a more commonly used term in the context of SDLC.

Reviewer: But if you feel strongly about "facet" that is fine... my main point is to be consistent. 👍

.. image:: images/design-dev-test.png
   :alt: Iterative feedback loop of design, development, and testing facets as described below.
   :align: center

> Note that the SDLC of a Daml application is not a linear process. In particular, the facets of design, develop, and test are not sequential but part of a dynamic, iterative process. These facets operate in a feedback loop, where the outcomes of one facet, such as testing, frequently lead back to adjustments in design or development. This iterative approach ensures continuous refinement and validation of the solution until it satisfies the required standards and objectives.

Reviewer: This iterative approach ensures continuous refinement and validation of the solution until it satisfies the requirements.

1. Design 
=========

1.1 Daml Schema for Workflows
-----------------------------
* Daml provides a schema language for defining workflows and state.
* The implementation of this schema should be a fundamental part of the design phase.

1.2 Benefits and Trade-Offs
---------------------------

1.2.1 Benefits and Functional Characteristics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  1. Synchronization of Application Workflow State Across Organizational Boundaries
     Traditional state synchronization relies on messaging, often requiring manual reconciliation. Daml simplifies this with:
     * A virtual shared ledger.
     * Built-in atomic transactions across organizations and applications.
  2. Functional Characteristics
     Using Daml and Canton to facilitate cross-organization workflows offers several unique advantages over using a database or other distributed ledger technologies. These include:
     * Non-repudiation: All transactions are traceable and cannot be denied by participants.

>      * Need-to-know privacy: Privacy is preserved by controlling data visibility to authorized parties.

Reviewer: I would move privacy to the top of the list whenever we are stating these benefits.

3. Test
=======

3.1 Testing Pyramid
-------------------
Testing Daml apps is similar to testing other systems: prioritize automation and test at the lowest level for speed and efficiency. However, building robust Daml apps comes with specific recommendations.

.. image:: images/testing-pyramid.png
   :alt: Testing pyramid as described below.
   :align: center

* Unit Tests
  * Use Daml Script for white-box unit tests.
  * Mock backends and ledgers for frontend testing.
* Integration Tests
  * Backend: Use white-box integration tests for internal APIs that are only used by clients under the app provider’s control.
  * Public APIs: Use black-box behavioral tests interacting at system boundaries.
  * Test isolation: Use long-running Canton instances to avoid repeatedly paying Canton’s startup cost, and isolate tests using unique participant users and parties for each test run. One approach is appending a test run ID as a suffix to party and user names in your test harness.
* End-to-End Tests
  * Test workflows between end-users and systems across multiple participant nodes, backends, and frontends.

>   * Use tools like Selenium or Playwright for browser session orchestration.

Reviewer: Helpful to link-out to Selenium and Playwright?

  * Test isolation: Either bootstrap the entire system for each test run or use a long-running system instance to specific tests. The latter approach supports faster test execution and quicker iterations.

3.2 Flaky Tests and Time Dependencies
-------------------------------------

> * Eliminate flaky tests to ensure developer productivity. Writing robust tests for distributed systems is challenging due to data propagation delays and concurrent execution. It is essential to identify and address flaky tests–those that fail inconsistently and incorrectly. 

Reviewer 2: ```suggestion
* Writing robust tests for distributed systems is challenging due to data propagation delays and concurrent execution. Eliminate flaky tests (those that fail inconsistently and incorrectly) to ensure developer productivity.  
```

> * For time-sensitive workflows, use the “passTime” function in Daml and configure reduced wait-times for faster CI execution. Workflows that incorporate calendar or time functions in their logic, such as bond lifecycling with coupon payments, can be tested by advancing time with “passTime.” For end-to-end tests, configure workflows to advance in milliseconds to reduce CI execution time. Pause and resume automation from the test harness to prevent race conditions.

Reviewer: `passTime` should be formatted in code ticks as opposed to quotes

> * For time-sensitive workflows, use the “passTime” function in Daml and configure reduced wait-times for faster CI execution. Workflows that incorporate calendar or time functions in their logic, such as bond lifecycling with coupon payments, can be tested by advancing time with “passTime.” For end-to-end tests, configure workflows to advance in milliseconds to reduce CI execution time. Pause and resume automation from the test harness to prevent race conditions.

Reviewer 2: ```suggestion
* For time-sensitive workflows, use the “passTime” function in Daml and configure reduced wait times for faster CI execution. Workflows with calendar or time functions in their logic, such as bond lifecycling with coupon payments, can be tested by advancing time with “passTime.” For end-to-end tests, configure workflows to advance in milliseconds to reduce CI execution time. Pause and resume automation from the test harness to prevent race conditions.
```

3.3 Performance Testing
-----------------------
* Start performance testing early and continuously.

> * Create separate performance tests for each relevant workflow. Incorporate additionally developed workflows into existing performance tests or create new ones specifically for these workflows.

Reviewer 2: ```suggestion
* Create separate performance tests for each relevant workflow. Incorporate newly developed workflows into existing performance tests or create tests specifically for these workflows.
```

* Test at scale with synthetic data resembling production characteristics.
* Measure performance characteristics and reset them between test runs to detect regressions.
* Perform soak testing with long-running deployments to detect bottlenecks.
* Set up alerting to monitor system failures, tuning it over time for optimal observability. Well-tuned alerts established during development can be reused in operations to detect system health issues.

1. Deploy
=========

4.1 Deployment Topology
-----------------------

> * The deployment topology depends on the :doc:`app architecture <daml-app-arch-design>` and requires a pre-configured Canton infrastructure, including app provider and app users at their respective administrative domains.

Reviewer 2: ```suggestion
* The deployment topology depends on the :doc:`app architecture <daml-app-arch-design>` and requires a pre-configured Canton infrastructure, including app provider and app users on their respective administrative domains.
```

> * Shared DAR files must be deployed on all Canton participant nodes. Daml code defines the API for state and workflows synchronized across participant nodes, similar to .proto files for a gRPC server shared with gRPC client developers. It is recommended to store Daml code in a separate repo from backend and frontend code and provide app user organizations with a tarball or read-only access to this repo. This allows organizations to review and build the code to ensure confidence in the behavior or the DAR file installed on their participant nodes.

Reviewer: `.proto` in code ticks?

4.2 Deployment Responsibilities
-------------------------------

> Some cross-organizational coordination is always required to deploy a Daml application. Each organization needs to set up the Canton infrastructure components and deploy the application components running within its administrative domain. Additionally, each organization must integrate its Canton participant nodes and the application components it deploys with its IAM.

Reviewer 2: ```suggestion
Some cross-organizational coordination is always required to deploy a Daml application. Each organization must set up the Canton infrastructure components and deploy the application components running within its administrative domain. Additionally, each organization must integrate its Canton participant nodes and the application components it deploys with its IAM.
```
Have we defined the IAM acronym?

* :ref:`App provider <app-provider>` should:

>   * Deploy backend and frontend.

Reviewer: Deploy _the_

>   * Configure frontend to integrate with IAM.

Reviewer: Configure _the_

* :ref:`App user <app-user>` should:
  * Deploy and configure the frontend to integrate with their own IAM, whether the frontend is developed by the app provider or the user themselves.

>   * Deploy backend if developed in-house.

Reviewer: Deploy _the_ backend...

4.3 Cross-Organizational Coordination
-------------------------------------
* Reduce deployment complexity by minimizing components required from app users.
* Balance deployment simplicity with :ref:`architectural trade-offs <properties-summary>`.

5. Operations Support
=====================

5.1 Logging and Monitoring
--------------------------
* Standard operational considerations for Daml apps include logging and monitoring.
  * Logging: Regularly review logs during development and testing, such as by capturing logs in CI runs and using them for debugging CI failures.

>   * Monitoring: Capture metrics for all components and display the golden signals–latency, traffic, errors, and saturation–on dashboards. Example dashboards for Canton components are available in `the documentation <https://docs.daml.com/canton/usermanual/monitoring.html?_gl=1*qdpp48*_gcl_au*MTQ0ODAwODc0MC4xNzM3NDQzODUw*_ga*NDg1MTgxODM0LjE3MjA2MjEzNDc.*_ga_GVK9ZHZSMR*MTczNzQ0Mzg1MC45OC4xLjE3Mzc0NDM5NjEuNjAuMC4w&_ga=2.19913016.1097062857.1737443850-485181834.1720621347#hands-on-with-the-daml-enterprise-observability-example>`_.

Reviewer 2: ```suggestion
  * Monitoring: Capture metrics for all components and display the golden signals – latency, traffic, errors, and saturation – on dashboards. Example dashboards for Canton components are available in `the documentation <https://docs.daml.com/canton/usermanual/monitoring.html?_gl=1*qdpp48*_gcl_au*MTQ0ODAwODc0MC4xNzM3NDQzODUw*_ga*NDg1MTgxODM0LjE3MjA2MjEzNDc.*_ga_GVK9ZHZSMR*MTczNzQ0Mzg1MC45OC4xLjE3Mzc0NDM5NjEuNjAuMC4w&_ga=2.19913016.1097062857.1737443850-485181834.1720621347#hands-on-with-the-daml-enterprise-observability-example>`_.
```

> * Set up alerting on the metrics to monitor the application’s health during testing and development. This ensures operational reuse and integration into the long-running test instance.

Reviewer: Should it be: "Set up _alerts_ ? Or is "alerting" a special key word in this context?

5.2 Upgrading
-------------
* Bug fixes and feature rollouts for off-ledger components follow standard design and development practices, similar to Web2 apps and enterprise services.
* Rolling out changes to Daml code requires additional considerations:
  * Daml code represents shared rules, requiring coordination across multiple administrative domains during upgrade.

>   * As an API definition for cross-organization workflows, changes to Daml code must be reflected in all components using it, ideally with backward compatibility to minimize code updates in dependent systems.

Reviewer 2: ```suggestion
  * As an API definition for cross-organization workflows, changes to Daml code must be reflected in all components using that code, ideally with backward compatibility to minimize code updates in the dependent systems.
```
